src/Firestore/PathTrait.php

```diff
@@ -0,0 +1,197 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Firestore;
+
+/**
+ * Provides methods for managing resource paths.
+ */
+trait PathTrait
+{
+    /**
+     * Create a full name from a project, database and relative path.
+     *
+     * @param string $projectId
+     * @param string $database
+     * @param string $relativeName
+     * @return string
+     */
+    private function fullName($projectId, $database, $relativeName)
+    {
+        $template = 'projects/%s/databases/%s/documents/%s';
+        return sprintf($template, $projectId, $database, $relativeName);
+    }
+
+    /**
+     * Create a full name from a database name and relative path.
+     *
+     * @param string $databaseName
+     * @param string $relativeName
+     * @return string
+     */
+    public function fullNameFromDatabase($databaseName, $relativeName)
+    {
+        $template = '%s/documents/%s';
+        return sprintf($template, $databaseName, $relativeName);
+    }
+
+    /**
+     * @param string $projectId The project ID.
+     * @param string $database The database name.
+     */
+    private function databaseName($projectId, $database)
+    {
+        return sprintf('projects/%s/databases/%s', $projectId, $database);
+    }
+
+    /**
+     * Get the database name from a path.
+     *
+     * @param string $name
+     * @return string
+     */
+    private function databaseFromName($name)
+    {
+        $parts = explode('/databases/', $name);
+        return $parts[0] . '/databases/' . explode('/', $parts[1])[0];
+    }
+
+    /**
+     * Determine whether the given path is a document.
+     *
+     * @param string $name
+     * @return bool
+     */
+    private function isDocument($name)
+    {
+        if (!$this->isRelative($name)) {
+            $name = $this->relativeName($name);
+        }
+
+        $parts = $this->splitName($name);
+        return count($parts) > 0 && count($parts) % 2 === 0;
+    }
+
+    /**
+     * Determine whether the given path is a collection.
+     *
+     * @param string $name
+     * @return bool
+     */
+    private function isCollection($name)
+    {
+        if (!$this->isRelative($name)) {
+            $name = $this->relativeName($name);
+        }
+
+        $parts = $this->splitName($name);
+        return count($parts) % 2 === 1;
+    }
+
+    /**
+     * Determine whether the given path is relative or absolute.
+     *
+     * @param string $name
+     * @return bool
+     */
+    private function isRelative($name)
+    {
+        $parts = $this->splitName($name);
+        return count($parts) > 0 && $parts[0] !== 'projects';
+    }
+
+    /**
+     * Split a path into pieces at the separator (`/`).
+     *
+     * @param string $name
+     * @return array
+     */
+    private function splitName($name)
+    {
+        return explode('/', trim($name, '/'));
```

Comment:
The backend actually supports spaces in document and collection IDs.

Reply:
The `trim` call here just removes leading and trailing slashes to keep empty elements out of the resulting array.

Without trim, you'd get something like this: 

```
php > $path = '/projects/foo/';
php > var_dump(explode('/', $path));
php shell code:1:
array(4) {
  [0] =>
  string(0) ""
  [1] =>
  string(8) "projects"
  [2] =>
  string(3) "foo"
  [3] =>
  string(0) ""
}
```

It does not remove spaces, when a character mask is given as the 2nd argument:
```
php > $str = ' foo ';
php > var_dump(trim($str, '/'));
php shell code:1:
string(5) " foo "
```

Reply:
That make sense. Thanks for explaining PHP to me :)